core/operation.ts

```diff
@@ -125,20 +136,33 @@ export class OperationContext {
   }
 
   public self(): string {
-    return this.resolve(this.operation.proto.name);
+    return this.resolve({
+      schema: this.operation.proto.target.schema,
+      name: this.operation.proto.target.name
+    });
   }
 
   public name(): string {
     return this.operation.proto.name;
   }
 
-  public ref(name: string) {
+  public ref(ref: Resolvable) {
+    const name = typeof ref === "object" ? ref.schema + "." + ref.name : ref;
     this.operation.dependencies(name);
-    return this.resolve(name);
+    return this.resolve(ref);
   }
 
-  public resolve(name: string) {
-    return this.operation.session.resolve(name);
+  public resolve(ref: Resolvable) {
+    if (this.hasOutput) {
+      return this.operation.session.resolve(ref);
+    } else {
+      this.operation.session.compileError(
```

Comment:
i think this is incorrect - operations should always be able to `resolve` anything. what's invalid is for somebody to `ref` or `resolve` an operation which has `hasOutput: false`. (which is checked in `session.ts)